server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -411,6 +411,30 @@ public void onFailure(Exception e) {
         return true;
     }
 
+    private static void onMappingUpdateFailure(
+        Exception e,
+        IndexShard primary,
+        boolean isDelete,
+        long version,
+        String id,
+        BulkPrimaryExecutionContext context,
+        UpdateHelper.Result updateResult
+    ) {
+        Engine.Result r = exceptionToResult(e, primary, isDelete, version, id);
+        if (context.getMappingUpdateRetryCounter() == 0) {
+            // retry all mapping update errors once
+            // the errors may be a result of a concurrent modification of the mapping
+            // for example, when adding a dynamic field under the premise that the field limit has not been reached, yet
+            // (see Dynamic.UNTIL_LIMIT)
+            // but the field limit has been reached by a another concurrent operation
+            // retrying once is enough to prevent the issue from happening again and avoids the risk of infinite retry loops
+            context.markOperationAsExecuted(r);
```

Comment:
This will set
```
String index = request.index();
executionResult = BulkItemResponse.failure(
                    current.id(),
                    docWriteRequest.opType(),
                    new BulkItemResponse.Failure(index, result.getId(), result.getFailure(), result.getSeqNo(), result.getTerm())
                );
```

But will then retry. My concern is that we retry and then fail for some other reason. Or the mapping update is due to something other than the dynamic update limit. 

I honestly don't know what happens in all these cases. Changes to this state machine makes be nervous.

Reply:
> My concern is that we retry and then fail for some other reason.

Why is that a concern? We'll only retry once, so there's no infinite loop if that's what you were concerned about. Or did you mean that the retry could fail for a transient reason and re-trying again would have fixed it? I think we can also live with that situation as that's also the case today where we never retry.

Or do you mean that we may fail for another reason and therefore the failure reason in the bulk response isn't right? I don't think that's the case as it would override the mutable `executionResult` when it retires. It will either be set to a success result or another failure result when it retries.

> Or the mapping update is due to something other than the dynamic update limit.

The only code path that leads to a mapping update here is if there are dynamic mapping updates during an index request. See also

https://github.com/elastic/elasticsearch/blob/2ac55b2f50500944fa1de849cb60bab2dede9dd1/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L971-L974

Retries aren't something completely new that I've introduced. While it's new to retry on mapping updates, we're already retrying on optimistic concurrency control failures ([`retry_on_conflict`](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html#docs-update-api-query-params)):

https://github.com/elastic/elasticsearch/blob/2ac55b2f50500944fa1de849cb60bab2dede9dd1/server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java#L426-L432

Reply:
As discussed with @original-brownbear, the retry now only happens if there were concurrent updates to the mapping.